### ISSUE TITLE
Tracking stats from dumpsys meminfo in perfetto

### ISSUE THREAD

**Issue by funfoolsuzi** [google/perfetto#534]
TLDR we want these following in perfetto for some reasons:
- collect some stats from `dumpsys meminfo` like lost RAM, dynamic kernel(in Used RAM field)
- a high polling rate like less than every 100ms.
- the timestamps of these data can be the same with other stats like pss, and some more other `/proc/` data sources. Assuming negligible latency among these pollings or we are ok with the inaccuracy caused by the latency.

The 1st problem is that `dumpsys meminfo` is slow and it can take 1 to 2 seconds. That rate is too slow. So we want to move the calculation of these interesting fields from [AndroidManagerService.java](https://cs.android.com/android/platform/superproject/+/main:frameworks/base/services/core/java/com/android/server/am/ActivityManagerService.java;l=11781) to somewhere in Perfetto. This may solves the problem of polling rate. But then the calculation of lost RAM requires all processes' PSS. And this becomes a metric that is based on both process stats and system stats. So this leads to the 2nd problem: it is unclear whether it will need to belong to `linux.process_stats` or `linux.sys_stats` config-wise.

As downstream of perfetto, I can try to emit track events somewhere in an ad-hoc manner without addressing these issues directly. However, I want to check if there is opportunity to upstream a clean solution that may be useful to other downstreams. So I wonder what does Perfetto team think about this and any suggestion is apprecaited :)

**Comment by LalitMaganti** [comment]
We'd be open to discussing an upstream solution but I think any collection of PSS should be ruled out. Developers reach for PSS because it appears to be the "easy single number" that they can track over time. However, from our significant experience with memory on Chrome and Android, we believe PSS to be a highly flawed metric which we've worked very hard to move away from.

An appropriate replacement would be "anon + swap" for the process level and tracking ion/dma allocations at the system level. While you lose the benefit of having "single number", there are significant gains in reproducibility, explainability and correctness with this approach.

Basically, memory is too complicated to be simplified to a single number and any attempts to do so will inevitably end in failure.

**Comment by funfoolsuzi** [comment]
Thanks for the suggestion. I will start to include this at some point.
Let's set PSS aside for now. I wonder what would be a good way to allow Perfetto collect process and system stats in a more genetic way and also allow metrics to be derived from these collected stats.
I wonder if we can have an alternative & more generic `probes_data_source` to collect both processes and system stats at a shared sample rate and allow user to configure some derived metrics through some config.
I want to check with Perfetto folks thoughts, before I start to explore this further.

**Comment by LalitMaganti** [comment]
On Wed, 19 Jul 2023, 09:05 X Li, ***@***.***> wrote:

> Thanks for the suggestion. I will start to include this at some point.
> Let's set PSS aside for now. I wonder what would be a good way to allow
> Perfetto collect process and system stats in a more genetic way and also
> allow metrics to be derived from these collected stats.
>
Not sure what value this adds over what we have today.

I wonder if we can have an alternative & more generic probes_data_source to
> collect both processes and system stats at a shared sample rate and allow
> user to configure some derived metrics through some config.
> I want to check with Perfetto folks thoughts, before I start to explore
> this further.
>
We're not really in the business of doing any aggregation on device tbh (at
least not in the way you are suggesting). That seems like a better fit for
the statsd service in Android instead imo.
